src/ast/json.rs

```diff
@@ -392,14 +410,26 @@ impl<'arena> Ast<'arena> for JSON<'arena> {
         }
     }
 
-    fn insert_chars(&self) -> Box<dyn Iterator<Item = char>> {
-        match self {
-            JSON::True | JSON::False | JSON::Null | JSON::Field(_) | JSON::Str(_) => {
-                Box::new(std::iter::empty())
+    fn valid_chars() -> Box<dyn Iterator<Item = char>> {
+        Self::all_object_chars()
+    }
+
+    fn is_valid_child(&self, index: usize, c: char) -> bool {
+        if self.holds_paired_child() {
+            match index {
+                0 => Self::valid_key_chars().any(|x| x == c),
+                _ => Self::valid_chars().any(|x| x == c),
             }
-            JSON::Object(_) | JSON::Array(_) => Self::all_object_chars(),
+        } else if self.holds_child() {
+            Self::valid_chars().any(|x| x == c)
+        } else {
+            false
         }
     }
```

Comment:
I think that this would be better (easier to read and modify) as a `match`:
```rust
fn is_valid_child(&self, index: usize, c: char) -> bool {
    match self {
        // Constant values can never have children
        JSON::True | JSON::False | ... => false,
        // Collections can take anything as children (except fields, which can't be created)
        JSON::Array(_) | JSON::Object(_) => true,
        // Fields must have strings as their first child
        JSON::Field(_) => if index == 0 { c == CHAR_STRING } else { true }
    }
}
```

This would also mean that the aux functions aren't needed.